### PR TITLE
Add nginx micro image

### PR DESCRIPTION
**Pull request by hhorak**
This is a PoC so far, meant to be available in quay.io initially, but ready to review and merge. The change adds a new version with a focus to deliver a container image with much smaller footprint.

It's supposed to be as small as possible, especially focusing on minimizing RPM dependencies. At the end though, the image does not miss many features comparing to the original one -- one significant change is a missing non-essential modules (nginx-mod-stream and nginx-mod-http-perl). Installing those images would add additional 55 MB uncompressed data, thanks to other dependencies it brings together.

### CODE REVIEW

**Comment by hhorak** [comment]
I've just learnt there is also some other attempt to make the nginx smaller, some WIP work visible at https://github.com/rexagod/base-ubi-nginx/tree/develop

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
The `run_dockerfiles_test` test is expected to fail, because it pulls in the remote github content that does not include `1.22-minimal` symlink yet.

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
The RHEL9 has some issues with subscription (being solved). Otherwise, it should be ready to review.

File: 1.22-minimal/Dockerfile.c9s [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+RUN mkdir -p /mnt/rootfs
+ADD https://copr.fedorainfracloud.org/coprs/hhorak/nginx-micro/repo/centos-stream-9/hhorak-nginx-micro-centos-stream-9.repo /etc/yum.repos.d/hhorak-nginx-micro-centos-stream-9.repo
+RUN MICRO_PKGS="coreutils-single glibc-minimal-langpack" && \
+    INSTALL_PKGS="$MICRO_PKGS nginx-core findutils hostname nss_wrapper-libs envsubst bind-utils" && \
+    dnf install --installroot /mnt/rootfs $INSTALL_PKGS --releasever 9 --setopt install_weak_deps=false --nodocs -y && \
+    dnf -y --installroot /mnt/rootfs clean all && \
+    rm -rf /mnt/rootfs/var/cache/* /mnt/rootfs/var/log/dnf* /mnt/rootfs/var/log/yum.*
+
+FROM quay.io/centos/centos:stream9
```
Comment (pkubatrh) [inline 1.22-minimal/Dockerfile.c9s:11]: Would use `FROM scratch` here, as there is no reason to layer it on top of a full sized image. We already have all we need from the previous build phase.

Comment (hhorak) [reply 1.22-minimal/Dockerfile.c9s:11]: will try, thanks for suggestion

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
It's now renamed to `micro` instead of `minimal`, to match naming in the httpd container. The idea is to use `-micro` when we use the two-stage build, which results in no rpm/yum in the resulting image. In contrary, `-minimal` will be based on base-minimal image and include at least `microdnf`.

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
[test-openshift]

**Comment by hhorak** [comment]
[test]

**Comment by phracek** [comment]
[test-all]

**Comment by phracek** [comment]
Let's test separate versions.

[test-all]

**Comment by hhorak** [comment]
Rebased and 1.20/Dockerfile.rhel9 fixed

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
[test-all]

**Review by phracek — CHANGES_REQUESTED** [review]
In order to test micro/minimal image please file another PR to test this image https://github.com/sclorg/nginx-container/blob/master/.github/workflows/container-tests.yml#L13 and https://github.com/sclorg/nginx-container/blob/master/.github/workflows/openshift-tests.yml#L13. Otherwise we do not now where micro/minimal image is working properly.

Also in case you are considering to push micro image to quay.io please add version here: https://github.com/sclorg/nginx-container/blob/master/.github/workflows/build-and-push.yml#L30

File: 1.22/Dockerfile.rhel8 [hunk 2]
```diff
@@ -45,10 +45,10 @@ RUN yum -y module enable nginx:$NGINX_VERSION && \
     yum -y clean all --enablerepo='*'
 
 # Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH
-COPY ./s2i/bin/ $STI_SCRIPTS_PATH
+COPY 1.22/s2i/bin/ $STI_SCRIPTS_PATH
```
Comment (zmiklank) [inline 1.22/Dockerfile.rhel8:48]: Why somewhere is used `$NGINX_VERSION` and somewhere a hardcoded version in these cases?

Comment (hhorak) [reply 1.22/Dockerfile.rhel8:48]: RHEL-8 somehow does not like it. I'm not sure whether it's a weird version of podman or what, but just simple change like this https://github.com/sclorg/nginx-container/pull/184/commits/7af4049fb857cccb674d00bda444795ddaf9c07a fixed the CI. This is still a bit mistery to me though.

File: 1.22-micro/Dockerfile.c9s [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+ENV NGINX_CONFIGURATION_PATH=${APP_ROOT}/etc/nginx.d \
+    NGINX_CONF_PATH=/etc/nginx/nginx.conf \
+    NGINX_DEFAULT_CONF_PATH=${APP_ROOT}/etc/nginx.default.d \
+    NGINX_CONTAINER_SCRIPTS_PATH=/usr/share/container-scripts/nginx \
+    NGINX_APP_ROOT=${APP_ROOT} \
+    NGINX_LOG_PATH=/var/log/nginx
+
+# Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH
+COPY $NGINX_VERSION-micro/s2i/bin/ $STI_SCRIPTS_PATH
```
[the remark this answers is not in the document]
Comment (hhorak) [reply 1.22-micro/Dockerfile.c9s:60]: Renamed after feedback from pkubat. We agreed on using the terminology that is "micro" for the two-stage images that do not have rpm/yum and "minimal" for the images having ubi-minimal parent (and having microdnf). Not sure where to document this background.

File: 1.22-micro/1.22-micro [hunk 4]
```diff
@@ -0,0 +1 @@
+.
```
[the remark this answers is not in the document]
Comment (hhorak) [reply 1.22-micro/1.22-micro]: Removed.

**Comment by hhorak** [comment]
> In order to test micro/minimal image please file another PR...

Thanks for the point, added.

**Comment by hhorak** [comment]
[test-all]

**Comment by hhorak** [comment]
It seems all points were addressed. Let's merge and polish potential deficiencies over time.

**Comment by phracek** [comment]
[test-all]

**Pull request merged**